service/controller/v12/resource/migration/resource.go

```diff
@@ -57,23 +61,38 @@ func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
 	if err != nil {
 		return microerror.Mask(err)
 	}
+	customObject = *customObject.DeepCopy()
 
-	// If credential secret is set we have nothing to migrate.
-	if customObject.Spec.AWS.CredentialSecret.Name != "" {
-		return nil
+	oldSpec := customObject.Spec.DeepCopy()
+
+	if reflect.DeepEqual(providerv1alpha1.AWSConfigSpecAWSHostedZones{}, customObject.Spec.AWS.HostedZones) {
+		r.logger.LogCtx(ctx, "level", "debug", "message", "CR is missing hosted zone names")
+
+		apiDomain := customObject.Spec.Cluster.Kubernetes.API.Domain
+		zone, err := zoneFromAPIDomain(apiDomain)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("settings all zones to %q", zone))
+
+		customObject.Spec.AWS.HostedZones.API.Name = zone
+		customObject.Spec.AWS.HostedZones.Etcd.Name = zone
+		customObject.Spec.AWS.HostedZones.Ingress.Name = zone
 	}
 
-	r.logger.LogCtx(ctx, "level", "debug", "message", "CR is missing credential, setting the default")
+	if reflect.DeepEqual(customObject.Spec, oldSpec) {
+		return nil
+	}
```

Comment:
I found this a nice way to quickly return when there aren't changes to migrate.
I think we should keep the migration resource even if there is nothing to migrate. Just to make it easier to migrate stuff if needed. I really like it.